fix(switchover): lazy shutdown and destroy

tiagolobocastro opened the pull request:
fix(shutdown): don't fail when lazily shutting down the nexus

When the core agent restarts at the same time as the older nexus node, then we don't have
the node state and as such we fail the shutdown call.
For a republish call we want to shutdown in a lazy manner, so ignore this error.
The reconcilers will pick up the slack, if need be.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

fix(switchover): destroy shutdown targets lazily

The cluster agent has some issues handling this itself, eg: it doesn't know when nodes are
online. It would have to ask the core-agent. Since it doesn't have this functionality yet
it's better to have the core agent handle these destroys lazily.

Also added a switchover test for this effect.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix(shutdown): don't fail when lazily shutting down the nexus

When the core agent restarts at the same time as the older nexus node, then we don't have
the node state and as such we fail the shutdown call.
For a republish call we want to shutdown in a lazy manner, so ignore this error.
The reconcilers will pick up the slack, if need be.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix(switchover): destroy shutdown targets lazily

The cluster agent has some issues handling this itself, eg: it doesn't know when nodes are
online. It would have to ask the core-agent. Since it doesn't have this functionality yet
it's better to have the core agent handle these destroys lazily.

Also added a switchover test for this effect.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

Abhinandan-Purkait commented on common/src/types/v0/store/mod.rs:
```diff
@@ -45,18 +45,26 @@ impl<T> From<SpecStatus<T>> for models::SpecStatus {
     pub fn deleted(&self) -> bool {
         self == &Self::Deleted
     }
+    /// Check if resource is being delete or is deleted.
```
`is being delete` -> `is being deleted`

Abhinandan-Purkait commented on common/src/types/v0/store/mod.rs:
```diff
@@ -45,18 +45,26 @@ impl<T> From<SpecStatus<T>> for models::SpecStatus {
     }
+    /// Check if resource is being delete or is deleted.
+    pub fn deleting_or_deleted(&self) -> bool {
+        self == &Self::Deleting
```
Should we also add `||  self == &Self::Deleted` ?

tiagolobocastro replied:
oops, well spotted!

Abhinandan-Purkait commented on common/src/types/v0/transport/nexus.rs:
```diff
@@ -409,8 +412,18 @@ impl DestroyNexus {
             node,
             uuid,
             disowners: NexusOwners::None,
+            lazy: false,
```
Should we take the lazy from parameters here as well like in ShutdownNexus below?

tiagolobocastro replied:
Most calls would set it to false anyway, which is why I didn't add it, it can be add with the builder type method.

tiagolobocastro committed:
fix(bdd): trap errors correctly

Make cleanup script executable and add trap before the tests are run otherwise we
hide execution errors and CI always passes even on failure..

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge
